### PR TITLE
Add TOTP as an authentication option

### PR DESCRIPTION
**Pull request by sampoder**
Screenshot 2024-06-06 at 12 04 31 PM

(dev data only)

Screenshot 2024-06-06 at 12 04 53 PM

### CODE REVIEW

File: app/models/user.rb [hunk 1]
```diff
@@ -162,7 +164,7 @@ class User < ApplicationRecord
     end
   end
 
-  scope :currently_online, -> { where(id: UserSession.where("last_seen_at > ?", 15.minutes.ago).pluck(:user_id)) }
+  scope :currently_online, -> { where(id: UserSession.where("created_at > ?", 15.minutes.ago).pluck(:user_id)) }
```
Comment (sampoder) [inline app/models/user.rb]:
```suggestion
  scope :currently_online, -> { where(id: UserSession.where("last_seen_at > ?", 15.minutes.ago).pluck(:user_id)) }
```

unrelated changes!

File: app/models/user.rb
Comment (garyhtou) [inline app/models/user.rb:1]: Please create a separate model for this! Here's why:

- https://blog.arkency.com/oop-refactoring-from-a-god-class-to-smaller-objects/
- https://refactoring.guru/smells/large-class

File: app/models/user.rb [hunk 2]
```diff
@@ -10,7 +10,9 @@
 #  comment_notifications    :integer          default("all_threads"), not null
 #  email                    :text
 #  full_name                :string
+#  last_otp_at              :datetime
 #  locked_at                :datetime
+#  otp_secret               :string
```
Comment (sampoder) [inline app/models/user.rb]:
```suggestion
#  locked_at                :datetime
```

File: db/schema.rb [hunk 3]
```diff
@@ -2010,6 +2019,8 @@
     t.string "payout_method_type"
     t.bigint "payout_method_id"
     t.integer "comment_notifications", default: 0, null: false
+    t.string "otp_secret"
+    t.datetime "last_otp_at"
```
Comment (sampoder) [inline db/schema.rb]:
```suggestion
```

**Comment by sampoder** [comment]
Yeah fs, just refactored it! Thoughts?

File: app/controllers/users_controller.rb [hunk 4]
```diff
@@ -169,6 +175,39 @@ def webauthn_auth
...
+    flash[:error] = "Please enter an email address."
+    redirect_to auth_users_path
+  end
+
+  def totp_auth
+    user = User.find_by(email: params[:email])
+
+    if !user
+      return redirect_to auth_users_path
+    end
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    return redirect_to auth_users_path unless user.present?
```

File: app/models/user/totp.rb [hunk 5]
```diff
@@ -0,0 +1,39 @@
+# frozen_string_literal: true
+
+# == Schema Information
+#
+# Table name: user_totps
+#
+#  id           :bigint           not null, primary key
+#  last_used_at :datetime
+#  secret       :string           not null
```
Comment (cjdenio) [inline app/models/user/totp.rb]: `secret` should be encrypted!

File: app/controllers/users_controller.rb [hunk 6]
```diff
@@ -169,6 +175,39 @@ def webauthn_auth
...
+
+    if user.totp.verify(params[:code], after: user.totp.last_used_at)
+      user.totp.update!(last_used_at: DateTime.now)
+      session[:login_preference] = "totp" if params[:remember] == "true"
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: is this line needed?

File: app/views/users/edit_security.html.erb [hunk 7]
```diff
@@ -9,35 +9,53 @@
...
-              </a>
-            </action>
-          </p>
-        <% end %>
-      </div>
+        </div>
+      <% end %>
     <% end %>
-  <% end %>
+    <div>
```
Comment (cjdenio) [inline app/views/users/edit_security.html.erb]: A few things:
- We should show the raw secret so you can set it up without pulling out your phone
- The secret should be only be accessible during setup
- There should be a way to re-generate the secret

File: app/models/user.rb [hunk 8]
```diff
@@ -309,6 +310,10 @@ def last_seen_at
     user_sessions.maximum(:last_seen_at)
   end
 
+  def totp
+    super || (build_totp.save && super)
```
Comment (garyhtou) [inline app/models/user.rb]:
```suggestion
    super || (create_totp && super)
```

File: app/models/user/totp.rb [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+#
+# Indexes
+#
+#  index_user_totps_on_user_id  (user_id)
+#
+class User
+  class Totp < ApplicationRecord
+    belongs_to :user
+    before_validation do
+      self.secret = ROTP::Base32.random
```
Comment (garyhtou) [inline app/models/user/totp.rb]:
```suggestion
      self.secret ||= ROTP::Base32.random
```

File: app/models/user/totp.rb [hunk 10]
```diff
@@ -0,0 +1,39 @@
...
+    delegate :verify, to: :instance
+
+    def provisioning_uri
+      instance.provisioning_uri(user.email)
+    end
+
+    private
+
+    def instance
+      ROTP::TOTP.new(secret, issuer: "hcb.hackclub.com")
```
Comment (garyhtou) [inline app/models/user/totp.rb]:
```suggestion
      ROTP::TOTP.new(secret, issuer: "HCB")
```

File: app/controllers/users_controller.rb [hunk 11]
```diff
@@ -169,6 +175,39 @@ def webauthn_auth
...
+  end
+
+  def totp_auth
+    user = User.find_by(email: params[:email])
+
+    if !user
+      return redirect_to auth_users_path
+    end
+
+    if user.totp.verify(params[:code], after: user.totp.last_used_at)
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]:
```suggestion
    if user.totp.verify(params[:code], drift_behind: 15, after: user.totp.last_used_at)
```

File: app/views/users/choose_login_preference.html.erb [hunk 12]
```diff
@@ -1,5 +1,6 @@
 <% title "Sign in" %>
-<% page_xs %>
+<% page_sm if @webauthn_available %>
+<% page_xs unless @webauthn_available %>
```
Comment (garyhtou) [inline app/views/users/choose_login_preference.html.erb]:
```suggestion
<% @webauthn_available ? page_sm : page_xs %>
```

File: app/models/user.rb (hunk 8 above)
Comment (cjdenio) [inline app/models/user.rb]:
```suggestion
    super || create_totp
```

File: app/views/users/edit_security.html.erb [hunk 13]
```diff
@@ -9,35 +9,53 @@
...
+      <legend class="flex flex-row justify-start items-center mt3 pb2">
+        <span class="h2">TOTP QR Code</span>
+        <%= badge_for "What's this?", class: "bg-muted" %>
+      </legend>
+      <qr-code
+        contents='<%= @user.totp.provisioning_uri %>'
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]:
```suggestion
        contents="<%= @user.totp.provisioning_uri %>"
```

File: app/views/users/totp.html.erb [hunk 14]
```diff
@@ -0,0 +1,26 @@
+<% title "Enter login code" %>
```
Comment (garyhtou) [inline app/views/users/totp.html.erb]:
```suggestion
<% title "Enter TOTP code" %>
```

File: app/views/users/edit_security.html.erb [hunk 15]
```diff
@@ -9,35 +9,53 @@
...
-          </p>
-        <% end %>
-      </div>
+        </div>
+      <% end %>
     <% end %>
-  <% end %>
+    <div>
+      <legend class="flex flex-row justify-start items-center mt3 pb2">
+        <span class="h2">TOTP QR Code</span>
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]: link change log post

File: app/models/user/totp.rb
Comment (garyhtou) [inline app/models/user/totp.rb:1]: Add soft deletion to this model

File: app/views/users/edit_security.html.erb (hunk 7 above)
Comment (sampoder) [reply app/views/users/edit_security.html.erb]: gotcha! thanks for all the review comments, I'll work on these today

File: app/views/users/edit_security.html.erb [hunk 16]
```diff
@@ -38,6 +38,25 @@
       </div>
     <% end %>
   <% end %>
+  <div>
+    <legend class="flex flex-row justify-start items-center mt3 pb2">
+      <span class="h2">Time-based one time passwords</span>
+      <%= badge_for "What's this?", class: "bg-muted" %>
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]: Change log post here

File: app/views/users/edit_security.html.erb [hunk 17]
```diff
@@ -38,6 +38,25 @@
...
+    </legend>
+    <%= turbo_frame_tag :totp_auth_card do %>
+      <div class="card border b--info mt1 mb3">
+        <% if @user.totp.present? %>
+          <p>
+            You set up time-based one time passwords on <%= @user.totp.created_at.strftime("%B %e, %Y") %>; if you'd like to regenerate your TOTP secret, <%= link_to "click here", user_enable_totp_path(@user), data: { turbo_method: :POST } %>. A new TOTP secret will invalidate your current TOTP secret.
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]:
```suggestion
            You set up time-based one time passwords on <%= @user.totp.created_at.strftime("%B %e, %Y") %>; if you'd like to reset your TOTP secret, <%= link_to "click here", user_enable_totp_path(@user), data: { turbo_method: :POST } %>. A new TOTP secret will invalidate your current TOTP secret.
```

File: app/views/users/edit_security.html.erb [hunk 18]
```diff
@@ -38,6 +38,25 @@
...
+          </p>
+        <% else %>
+          <p>
+            You currently don't have time-based one time passwords setup, <%= link_to "click here", user_enable_totp_path(@user), data: { turbo_method: :POST } %> to setup TOTPs.
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]: set up? setup?

Comment (garyhtou) [inline app/views/users/edit_security.html.erb]:
```suggestion
            You currently don't have time-based one time passwords setup, <%= link_to "click here", user_enable_totp_path(@user), data: { turbo_method: :POST } %> to setup your TOTP.
```

File: app/views/users/enable_totp.html.erb [hunk 19]
```diff
@@ -0,0 +1,19 @@
+<%= turbo_frame_tag :totp_auth_card do %>
+  <div class="card border b--info mt1 mb3">
+    <p>
+      Scan the QR code below with your authenticator app or enter the secret.
+      FYI: you can only see this once! After that, you'll have to set up a new TOTP secret.
```
Comment (garyhtou) [inline app/views/users/enable_totp.html.erb]:
```suggestion
      FYI: you can only see this once! When you leave the page without configuring your authenticator app, you will need to reset your TOPT secret.
```

File: db/migrate/20240606180502_create_user_totp.rb [hunk 20]
```diff
@@ -0,0 +1,11 @@
+class CreateUserTotp < ActiveRecord::Migration[7.1]
+  def change
+    create_table :user_totps do |t|
+      t.belongs_to :user, null: false, index: true
+      t.string :secret, null: false
```
Comment (garyhtou) [inline db/migrate/20240606180502_create_user_totp.rb]:
```suggestion
```

File: app/models/user/totp.rb [hunk 21]
```diff
@@ -0,0 +1,40 @@
...
+#  user_id           :bigint           not null
+#
+# Indexes
+#
+#  index_user_totps_on_user_id  (user_id)
+#
+class User
+  class Totp < ApplicationRecord
+    belongs_to :user
+    has_encrypted :secret
```
Comment (garyhtou) [inline app/models/user/totp.rb]:
```suggestion
    has_encrypted :secret
    validates :secret, presence: true
```

File: app/controllers/users_controller.rb [hunk 22]
```diff
@@ -326,6 +366,13 @@ def edit_security
     authorize @user
   end
 
+  def enable_totp
+    @user = params[:id] ? User.friendly.find(params[:id]) : current_user
+    authorize @user
+    @user.totp&.destroy!
+    @user.create_totp
```
Comment (garyhtou) [inline app/controllers/users_controller.rb]: test this!
```suggestion
    @user.create_totp!
```

File: app/views/users/edit_security.html.erb [hunk 23]
```diff
@@ -38,6 +38,25 @@
       </div>
     <% end %>
   <% end %>
+  <div>
```
Comment (garyhtou) [inline app/views/users/edit_security.html.erb]: Feature flag this

**Review by garyhtou — COMMENTED** [review]
🪄
